packages/mui-base/src/TextareaAutosize/TextareaAutosize.js

```diff
@@ -86,41 +91,72 @@ const TextareaAutosize = React.forwardRef(function TextareaAutosize(props, ref)
     const outerHeightStyle = outerHeight + (boxSizing === 'border-box' ? padding + border : 0);
     const overflow = Math.abs(outerHeight - innerHeight) <= 1;
 
-    setState((prevState) => {
-      // Need a large enough difference to update the height.
-      // This prevents infinite rendering loop.
-      if (
-        renders.current < 20 &&
-        ((outerHeightStyle > 0 &&
-          Math.abs((prevState.outerHeightStyle || 0) - outerHeightStyle) > 1) ||
-          prevState.overflow !== overflow)
-      ) {
-        renders.current += 1;
-        return {
-          overflow,
-          outerHeightStyle,
-        };
-      }
+    return { outerHeightStyle, overflow };
+  }, [maxRows, minRows, props.placeholder]);
 
-      if (process.env.NODE_ENV !== 'production') {
-        if (renders.current === 20) {
-          console.error(
-            [
-              'MUI: Too many re-renders. The layout is unstable.',
-              'TextareaAutosize limits the number of renders to prevent an infinite loop.',
-            ].join('\n'),
-          );
-        }
+  const updateState = (prevState, newState) => {
+    const { outerHeightStyle, overflow } = newState;
+    // Need a large enough difference to update the height.
+    // This prevents infinite rendering loop.
+    if (
+      renders.current < 20 &&
+      ((outerHeightStyle > 0 &&
+        Math.abs((prevState.outerHeightStyle || 0) - outerHeightStyle) > 1) ||
+        prevState.overflow !== overflow)
+    ) {
+      renders.current += 1;
+      return {
+        overflow,
+        outerHeightStyle,
+      };
+    }
+    if (process.env.NODE_ENV !== 'production') {
+      if (renders.current === 20) {
+        console.error(
+          [
+            'MUI: Too many re-renders. The layout is unstable.',
+            'TextareaAutosize limits the number of renders to prevent an infinite loop.',
+          ].join('\n'),
+        );
       }
+    }
+    return prevState;
+  };
+
+  const syncHeight = React.useCallback(() => {
+    const newState = getUpdatedState();
+
+    if (isEmpty(newState)) {
+      return;
+    }
 
-      return prevState;
+    setState((prevState) => {
+      return updateState(prevState, newState);
     });
-  }, [maxRows, minRows, props.placeholder]);
+  }, [getUpdatedState]);
+
+  const syncHeightWithFlushSycn = () => {
+    const newState = getUpdatedState();
+
+    if (isEmpty(newState)) {
+      return;
+    }
+
+    // In React 18, state updates in a ResizeObserver's callback are happening after the paint which causes flickering
```

Comment:
I know you're nerd-sniping me because I'm pretty sure this should be fixed with some experimental features 😛 On vacation now but I'll definitely take a look because having to reach for flushSync every time you're in a ResizeObserver callback seems odd and something you often miss unless you've encountered this bug.

Reply:
Haha it's not intentional, but yes, I agree it's strange and should be fixed once someone is back from vacation 😁 Just kidding, but really looks like a strange requirement and more importanly is not easy to capture the bugs or create reliable tests.